Error while checking build/O2/fullCI for 67e588788f04e40b8311a349ee801bcc74dd446a at 2023-08-07 14:43:
```
## sw/BUILD/O2Physics-latest/log
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11731/67e588788f04e40b8311a349ee801bcc74dd446a/build_O2_fullCI/pretty.html).

--- a/Detectors/TPC/calibration/include/TPCCalibration/TPCFastSpaceChargeCorrectionHelper.h
+++ b/Detectors/TPC/calibration/include/TPCCalibration/TPCFastSpaceChargeCorrectionHelper.h
@@ -86,8 +86,7 @@ class TPCFastSpaceChargeCorrectionHelper
 
   /// Create SpaceCharge correction out of the voxel tree
   std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> createFromTrackResiduals(
-    const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree);
-
+    const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree, bool useSmoothed = false, bool invertSigns = false);
   /// _______________  Utilities   ________________________
 
   const TPCFastTransformGeo& getGeometry() { return mGeo; }

--- a/Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx
+++ b/Detectors/TPC/calibration/src/TPCFastSpaceChargeCorrectionHelper.cxx
@@ -375,7 +375,7 @@ void TPCFastSpaceChargeCorrectionHelper::testGeometry(const TPCFastTransformGeo&
 }
 
 std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> TPCFastSpaceChargeCorrectionHelper::createFromTrackResiduals(
-  const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree)
+  const o2::tpc::TrackResiduals& trackResiduals, TTree* voxResTree, bool useSmoothed, bool invertSigns)
 {
   // create o2::gpu::TPCFastSpaceChargeCorrection  from o2::tpc::TrackResiduals::VoxRes voxel tree
 
@@ -502,11 +502,14 @@ std::unique_ptr<o2::gpu::TPCFastSpaceChargeCorrection> TPCFastSpaceChargeCorrect
     double dy = x / trackResiduals.getDY2XI(xBin, y2xBin);
     double dz = x * trackResiduals.getDZ2X(z2xBin);
 
-    double correctionX = v->DS[o2::tpc::TrackResiduals::ResX];
-    double correctionY = v->DS[o2::tpc::TrackResiduals::ResY];
-    double correctionZ = v->DS[o2::tpc::TrackResiduals::ResZ];
-    double correctionD = v->DS[o2::tpc::TrackResiduals::ResD];
-
+    double correctionX = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResX] : v->D[o2::tpc::TrackResiduals::ResX];
+    double correctionY = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResY] : v->D[o2::tpc::TrackResiduals::ResY];
+    double correctionZ = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResZ] : v->D[o2::tpc::TrackResiduals::ResZ];
+    if (invertSigns) {
+      correctionX *= -1.;
+      correctionY *= -1.;
+      correctionZ *= -1.;
+    }
     // add one point per voxel
 
     // map.addCorrectionPoint(iRoc, iRow, y, z, correctionX, correctionY,

--- a/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.cxx
+++ b/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.cxx
@@ -25,6 +25,10 @@
 
 using namespace GPUCA_NAMESPACE::gpu;
 
+#ifndef GPUCA_ALIROOT_LIB
+ClassImp(TPCFastSpaceChargeCorrection);
+#endif
+
 TPCFastSpaceChargeCorrection::TPCFastSpaceChargeCorrection()
   : FlatObject(),
     mConstructionRowInfos(nullptr),

--- a/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h
+++ b/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h
@@ -44,6 +44,9 @@ class TPCFastSpaceChargeCorrection : public FlatObject
   struct RowInfo {
     int splineScenarioID{0};      ///< scenario index (which of Spline2D splines to use)
     size_t dataOffsetBytes[3]{0}; ///< offset for the spline data withing a TPC slice
+#ifndef GPUCA_ALIROOT_LIB
+    ClassDefNV(RowInfo, 1);
+#endif
   };
 
   struct RowActiveArea {
@@ -52,19 +55,28 @@ class TPCFastSpaceChargeCorrection : public FlatObject
     float cuMin{0.};
     float cuMax{0.};
     float cvMax{0.};
+#ifndef GPUCA_ALIROOT_LIB
+    ClassDefNV(RowActiveArea, 1);
+#endif
   };
 
   struct SliceRowInfo {
-    float gridV0{0.};           // V coordinate of the V-grid start
-    float gridCorrU0{0.};       // U coordinate of the U-grid start for corrected U
-    float gridCorrV0{0.};       // V coordinate of the V-grid start for corrected V
-    float scaleCorrUtoGrid{0.}; // scale corrected U to U-grid coordinate
-    float scaleCorrVtoGrid{0.}; // scale corrected V to V-grid coordinate
+    float gridV0{0.};           ///< V coordinate of the V-grid start
+    float gridCorrU0{0.};       ///< U coordinate of the U-grid start for corrected U
+    float gridCorrV0{0.};       ///< V coordinate of the V-grid start for corrected V
+    float scaleCorrUtoGrid{0.}; ///< scale corrected U to U-grid coordinate
+    float scaleCorrVtoGrid{0.}; ///< scale corrected V to V-grid coordinate
     RowActiveArea activeArea;
+#ifndef GPUCA_ALIROOT_LIB
+    ClassDefNV(SliceRowInfo, 1);
+#endif
   };
 
   struct SliceInfo {
-    float vMax{0.};
+    float vMax{0.}; ///< Max value of V coordinate
+#ifndef GPUCA_ALIROOT_LIB
+    ClassDefNV(SliceInfo, 1);
+#endif
   };
 
   typedef Spline2D<float, 3> SplineType;
@@ -254,8 +266,6 @@ class TPCFastSpaceChargeCorrection : public FlatObject
   RowInfo* mRowInfoPtr;           //! (transient!!) pointer to RowInfo array inside the mFlatBufferPtr buffer
   SliceRowInfo* mSliceRowInfoPtr; //! (transient!!) pointer to SliceRowInfo array inside the mFlatBufferPtr
 
-  float fInterpolationSafetyMargin{0.1}; // 10% area around the TPC row. Outside of this area the interpolation returns the boundary values.
-
   /// _______________  Calibration data  _______________________________________________
 
   long int mTimeStamp; ///< time stamp of the current calibration
@@ -263,6 +273,9 @@ class TPCFastSpaceChargeCorrection : public FlatObject
   char* mSplineData[3]; //! (transient!!) pointer to the spline data in the flat buffer
 
   size_t mSliceDataSizeBytes[3]; ///< size of the data for one slice in the flat buffer
+
+  float fInterpolationSafetyMargin{0.1}; // 10% area around the TPC row. Outside of this area the interpolation returns the boundary values.
+
 #ifndef GPUCA_ALIROOT_LIB
   ClassDefNV(TPCFastSpaceChargeCorrection, 3);
 #endif

--- a/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrectionMap.cxx
+++ b/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrectionMap.cxx
@@ -16,3 +16,9 @@
 
 #include "TPCFastSpaceChargeCorrectionMap.h"
 #include "GPUCommonLogger.h"
+
+using namespace GPUCA_NAMESPACE::gpu;
+
+#ifndef GPUCA_ALIROOT_LIB
+ClassImp(TPCFastSpaceChargeCorrectionMap);
+#endif

--- a/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrectionMap.h
+++ b/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrectionMap.h
@@ -18,6 +18,7 @@
 #define ALICEO2_GPUCOMMON_TPCFASTTRANSFORMATION_TPCFASTSPACECHARGECORRECTIONMAP_H
 
 #include "GPUCommonDef.h"
+#include "Rtypes.h"
 #include <vector>
 
 namespace GPUCA_NAMESPACE
@@ -26,10 +27,11 @@ namespace gpu
 {
 
 ///
-/// The TPCFastSpaceChargeCorrectionMap class represents correction of nominal coordinates of TPC clusters
-/// using best-fit splines
+/// The TPCFastSpaceChargeCorrectionMap class represents a maps of corrections of nominal coordinates of TPC clusters
 ///
 /// Row, U, V -> dX,dU,dV
+///
+/// It is used by TPCFastSpoaceChargeCorrectionHelper
 ///
 /// The class is flat C structure. No virtual methods, no ROOT types are used.
 ///
@@ -94,6 +96,10 @@ class TPCFastSpaceChargeCorrectionMap
   int mNrocs{0};
   int mNrows{0};
   std::vector<std::vector<CorrectionPoint>> fDataPoints; //! (transient!!) points with space charge correction
+
+#ifndef GPUCA_ALIROOT_LIB
+  ClassDefNV(TPCFastSpaceChargeCorrectionMap, 0);
+#endif
 };
 
 } // namespace gpu

--- a/GPU/TPCFastTransformation/TPCFastTransformGeo.h
+++ b/GPU/TPCFastTransformation/TPCFastTransformGeo.h
@@ -20,6 +20,7 @@
 #include "GPUCommonDef.h"
 #ifndef GPUCA_GPUCODE_DEVICE
 #include <memory>
+#include "Rtypes.h"
 #endif
 
 namespace GPUCA_NAMESPACE
@@ -37,6 +38,9 @@ class TPCFastTransformGeo
   struct SliceInfo {
     float sinAlpha;
     float cosAlpha;
+#ifndef GPUCA_ALIROOT_LIB
+    ClassDefNV(SliceInfo, 1);
+#endif
   };
 
   /// The struct contains necessary info about TPC padrow
@@ -50,6 +54,9 @@ class TPCFastTransformGeo
 
     /// get width in U
     GPUd() float getUwidth() const { return -2.f * u0; }
+#ifndef GPUCA_ALIROOT_LIB
+    ClassDefNV(RowInfo, 1);
+#endif
   };
 
   /// _____________  Constructors / destructors __________________________
@@ -201,6 +208,10 @@ class TPCFastTransformGeo
 
   SliceInfo mSliceInfos[NumberOfSlices + 1]; ///< array of slice information [fixed size]
   RowInfo mRowInfos[MaxNumberOfRows + 1];    ///< array of row information [fixed size]
+
+#ifndef GPUCA_ALIROOT_LIB
+  ClassDefNV(TPCFastTransformGeo, 1);
+#endif
 };
 
 // =======================================================================

--- a/GPU/TPCFastTransformation/TPCFastTransformationLinkDef_O2.h
+++ b/GPU/TPCFastTransformation/TPCFastTransformationLinkDef_O2.h
@@ -57,14 +57,21 @@
 #pragma link C++ class o2::gpu::IrregularSpline2D3D + ;
 #pragma link C++ class o2::gpu::SemiregularSpline2D3D + ;
 #pragma link C++ class o2::gpu::IrregularSpline2D3DCalibrator + ;
+
 #pragma link C++ class o2::gpu::TPCFastTransformGeo + ;
 #pragma link C++ class o2::gpu::TPCFastTransformGeo::SliceInfo + ;
 #pragma link C++ class o2::gpu::TPCFastTransformGeo::RowInfo + ;
+
 #pragma link C++ class o2::gpu::TPCFastTransform + ;
+
 #pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrectionMap + ;
-#pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrectionMap::CorrectionPoint + ;
-#pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrection::SliceInfo + ;
+#pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrection::RowInfo + ;
+
 #pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrection + ;
+#pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrection::SliceInfo + ;
+#pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrection::RowActiveArea + ;
+#pragma link C++ class o2::gpu::TPCFastSpaceChargeCorrection::SliceRowInfo + ;
+
 #pragma link C++ class o2::gpu::CorrectionMapsHelper + ;
 #pragma link C++ struct o2::gpu::MultivariatePolynomialContainer + ;
 #pragma link C++ struct o2::gpu::NDPiecewisePolynomialContainer + ;

--- a/GPU/TPCFastTransformation/macro/TPCFastTransformInit.C
+++ b/GPU/TPCFastTransformation/macro/TPCFastTransformInit.C
@@ -40,7 +40,7 @@ using namespace o2::tpc;
 using namespace o2::gpu;
 
 void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
-                          const char* outFileName = "TPCFastTransform_VoxRes.root")
+                          const char* outFileName = "TPCFastTransform_VoxRes.root", bool useSmoothed = false, bool invertSigns = false)
 {
 
   // Initialise TPCFastTransform object from "voxRes" tree of
@@ -84,7 +84,7 @@ void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
 
   o2::tpc::TPCFastSpaceChargeCorrectionHelper* corrHelper = o2::tpc::TPCFastSpaceChargeCorrectionHelper::instance();
 
-  auto corrPtr = corrHelper->createFromTrackResiduals(trackResiduals, voxResTree);
+  auto corrPtr = corrHelper->createFromTrackResiduals(trackResiduals, voxResTree, useSmoothed, invertSigns);
 
   std::unique_ptr<o2::gpu::TPCFastTransform> fastTransform(
     helper->create(0, *corrPtr));
@@ -195,10 +195,15 @@ void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
       z = -z;
     }
 
-    double correctionX = v->D[o2::tpc::TrackResiduals::ResX];
-    double correctionY = v->D[o2::tpc::TrackResiduals::ResY];
-    double correctionZ = v->D[o2::tpc::TrackResiduals::ResZ];
-    double correctionD = v->D[o2::tpc::TrackResiduals::ResD];
+    double correctionX = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResX] : v->D[o2::tpc::TrackResiduals::ResX];
+    double correctionY = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResY] : v->D[o2::tpc::TrackResiduals::ResY];
+    double correctionZ = useSmoothed ? v->DS[o2::tpc::TrackResiduals::ResZ] : v->D[o2::tpc::TrackResiduals::ResZ];
+
+    if (invertSigns) {
+      correctionX *= -1.;
+      correctionY *= -1.;
+      correctionZ *= -1.;
+    }
 
     // TODO: skip empty voxels?
     if (voxEntries < 1.) { // no statistics

